odoo/models.py

```diff
@@ -1731,6 +1738,310 @@ def clear_caches(cls):
         """
         cls.pool._clear_cache()
 
+    @api.model
+    def _read_group_select(self, query, aggregate_spec):
+        """ return (<SQL expression>, [<fields name used in SQL expression>])"""
+        if aggregate_spec == '__count':
+            return 'COUNT(*)', []
+
+        fname, func = split_read_group_spec(aggregate_spec)
+
+        if fname not in self:
+            raise ValueError(f"Invalid field {fname!r} on model {self._name!r} for {aggregate_spec!r}.")
+        if func not in VALID_AGGREGATE_FUNCTIONS:
+            raise ValueError(f"Invalid aggregate method {func!r} for {aggregate_spec!r}.")
+
+        field_expression = self._inherits_join_calc(self._table, fname, query)
+
+        func = func.upper()
+        distinct = ''
+        if func.endswith('_DISTINCT'):
+            distinct = 'DISTINCT '
+            func = func[:-len('_DISTINCT')]
+
+        if func == 'ARRAY_AGG' and not distinct:
+            # ORDER BY to force the determinism of result for array_agg
+            # array_agg_distinct doesn't have this particularity because we cannot order on distinct aggregate:
+            # in an aggregate with DISTINCT, ORDER BY expressions must appear in argument list
+            expression = f'ARRAY_AGG({field_expression} ORDER BY {self._inherits_join_calc(self._table, "id", query)})'
+        else:
+            expression = f'{func}({distinct}{field_expression})'
+
+        return expression, [fname]
+
+    @api.model
+    def _read_group_groupby(self, query, groupby_spec):
+        """ return (<SQL expression>, [<fields name used in SQL expression>])"""
+        res_match = regex_aggregate_spec.match(groupby_spec)
+        if not res_match:
+            raise ValueError(f"Invalid groupby specification {groupby_spec!r}.")
+
+        fname, granularity = res_match.groups()
+        if fname not in self:
+            raise ValueError(f"Invalid field {fname!r} on model {self._name!r}")
+
+        field = self._fields[fname]
+
+        if granularity and field.type not in ('datetime', 'date'):
+            raise ValueError(f"Granularity set on a no-datetime field: {groupby_spec!r}")
+
+        if not field.store and not field.inherited:  # TODO: should be manage by _inherits_join_calc
+            raise ValueError(f"Groupby on no-store field: {groupby_spec!r}")
+
+        field_expression = self._inherits_join_calc(self._table, fname, query)
+        if field.type == 'datetime' and self.env.context.get('tz') in pytz.all_timezones_set:
+            # TODO: '{self._context.get('tz')}' should be passed as args
```

Comment:
if we still use tz from context, can we add an assert with a regex ? (in case one day they update the lib and we have a timezone like `Los Angele's` with a single quote)

Reply:
I prefer not, I don't like regex. We will create a compositional SQL String, it will solve this issue cleanly.